### ISSUE TITLE
freezed script

### ISSUE THREAD

**Issue by Trahtenberg** [puppeteer/puppeteer#4029]
Hi!

Is it possible to somehow stop (set a pause) the script. For example, when a modal window appears, stop the script execution, while closing the window, then defrost the script. I try this:
```

await page.waitForSelector('div[class="popup _active"]').then(async()=>{

	console.log('catch popup');

	page.click('span[class="submit"]').then(()=>{
		console.log('click confirm');
	});

	await page.waitForSelector('div[class="popup _active"]',{hidden:true}).then(()=>{
		console.log('el hidden');
	});
});

```
But it does not freeze script.

Is it possible to implement without using flags to make the script more flexible?

**Comment by vsemozhetbyt** [comment]
I would suggest to not mix `async/await` with `then` so that the flow would not be overcomplicated and some waiting would not be missed. Try this:
```js
await page.waitForSelector('div[class="popup _active"]');
console.log('catch popup');

await page.click('span[class="submit"]');
console.log('click confirm');

await page.waitForSelector('div[class="popup _active"]', { hidden: true });
console.log('el hidden');
```

**Comment by Trahtenberg** [comment]
Yes, it works, I tried to do more encapsulation, more versatility in the code. I took your advice, thank you for the answer!

**Comment by vsemozhetbyt** [comment]
If you really need the mixing approach, try to add missing `await` here in this way:
```js
	await page.click('span[class="submit"]').then(()=>{
		console.log('click confirm');
	});
```

**Comment by Trahtenberg** [comment]
OK, I understand, I will experiment.)))

**Comment by Trahtenberg** [comment]
I found and played with the method `Promise.all();`  but I have a new difficulty. 
Using this set:

```
await Promise.all([		
        page.click('a[class="next_page]'),
        page.waitForNavigation({waitUntil: ['networkidle0']}),
]);
```

This script is used at the end of the cycle to turn the pages of the site. But there is a premature button press. The page does not have time to load, the button is not yet active. On each page, the cycle is triggered several times. To solve this problem, I have to add `await page.waitFor(5000);` before block promises, which is not very convenient. If there are problems with the connection, the code will not work correctly. I don't understand which selector the script should be attached to so that it does not work sooner or later than the time assigned to it.

**Comment by vsemozhetbyt** [comment]
I am not sure I understand the issue. Can you provide the URL?

**Comment by Trahtenberg** [comment]
https://repl.it/@RomanRomanovich/puppeteer

**Comment by aslushnikov** [comment]
Hey @Trahtenberg, what exactly are you trying to do and what doesn't work? Can you please narrow down the scope as much as possible - it'll help us to understand your issue and help you.

For example, "I'm trying to click that big green button at this URL" is way easier than reverse-engineering your logic in 100-lines script.

**Comment by vsemozhetbyt** [comment]
The clicked element is not a button, it is a link. How a link can be inactive? Has it no `href` at this time?

**Comment by Trahtenberg** [comment]
I'll explain now. In the block where the button is pressed further, the waiting time is set. Because when the cycle starts again, the page does not have time to load and the condition of pressing the button works again (ie, the product is parsed again from the first page). You can comment out `waitForNavigation` because if you leave only it, the script hangs after clicking on the  button next, and `waitForNavigation` does not see the page load. If there are lite problems with the connection, the script will not work correctly. Now the question itself is how to organize the script in such a way that, using the minimum,` waitFor(seconds)`?

**Comment by aslushnikov** [comment]
> I'll explain now. In the block where the button is pressed further, the waiting time is set. Because when the cycle starts again, the page does not have time to load and the condition of pressing the button works again (ie, the product is parsed again from the first page). You can comment out waitForNavigation because if you leave only it, the script hangs after clicking on the button next, and waitForNavigation does not see the page load.

@Trahtenberg I'd appreciate an explanation in layman terms that doesn't refer to code - to help us understand what's going on.

> Now the question itself is how to organize the script in such a way that, using the minimum, waitFor(seconds)?

if you're testing your own website, then I'd recommend set up certain hooks on the page that help you test it. The usual practice is to use data attributes on elements and then wait/query for them with [`page.waitForSelector`](https://pptr.dev/#?product=Puppeteer&version=v1.12.2&show=api-pagewaitforselectorselector-options)/[`page.$`](https://pptr.dev/#?product=Puppeteer&version=v1.12.2&show=api-pageselector). Other nice approach is to use `console.timeStamp` to yield certain events - you can listen to them on the puppeteer side using [`page.on('metrics')`](https://pptr.dev/#?product=Puppeteer&version=v1.12.2&show=api-event-metrics) event.

If you're dealing with a website as a black box, then there's no silver bullet and all solutions vary a lot in creativity. Sometimes tracking certain network requests using [`page.on('requestfinished')`](https://pptr.dev/#?product=Puppeteer&version=v1.12.2&show=api-event-requestfinished) works nicely, sometimes there's a specific DOM element that you can wait for.

**Comment by Trahtenberg** [comment]
Thank you very much, I will try.

**Comment by aslushnikov** [comment]
Sure, I'll close this for now. Feel free to re-file with more details if you continue to experience issues.
